### PR TITLE
Fix failing tests after Logstash 8 support microseconds

### LINKED ISSUE #123: Implement IPFIX reduced size encoding support (Was: Template length exceeds flowset length)
In logs i see huge repeatedly messages:
`[WARN ][logstash.outputs.elasticsearch] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"elastiflow-2018.02.18", :_type=>"doc", :_routing=>nil}, #<LogStash::Event:0x73a7b610>], :response=>{"index"=>{"_index"=>"elastiflow-2018.02.18", "_type"=>"doc", "_id"=>"MByGp2EBu08NUWC7JEYT", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"failed to parse [netflow.tcpOptions]", "caused_by"=>{"type"=>"json_parse_exception", "reason"=>"Numeric value (17293822569104482748) out of range of long (-9223372036854775808 - 9223372036854775807)\n at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@119006d0; line: 1, column: 1761]"}}}}}` 

`[WARN ][logstash.codecs.netflow  ] Template length exceeds flowset length, skipping {:template_id=>260, :template_length=>65, :record_length=>64}`

and somewhat less often this messages:
`[WARN ][logstash.codecs.netflow  ] Received template 260 of size 61 bytes doesn't match BinData representation we built (65 bytes)`
`[WARN ][logstash.codecs.netflow  ] Received template 263 of size 49 bytes doesn't match BinData representation we built (53 bytes)`
`[WARN ][logstash.codecs.netflow  ] Template length exceeds flowset length, skipping {:template_id=>263, :template_length=>53, :record_length=>52}`

- logstash 6.2.1 / logstash-codec-netflow (3.11.2):
- Debian 8.10 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-2+deb8u3 (2016-07-02):
- Device ipt_NETFLOW master branch (https://github.com/aabc/ipt-netflow)
- Run Logstash and view the log file

Tshark pcap dump in attach.
Thanks a lot!
[tshark.zip](https://github.com/logstash-plugins/logstash-codec-netflow/files/1734440/tshark.zip)

### LINKED ISSUE #203: Fail test on CI for stack version 8.x
**Build scan**:
https://app.travis-ci.com/github/logstash-plugins/logstash-codec-netflow/jobs/613875530

**Repro line**:
```
export LOGSTASH_PATH=/tmp/logstash && export LOGSTASH_SOURCE=1
bundle install
bundle exec rspec
```

**Reproduces locally?**:
yes

**Applicable branches**:
`main`

**Failure history**:

**Failure excerpt**:
```
logstash_1_2681004e5afa | Failures:
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   1) LogStash::Codecs::Netflow configured with template caching IPFIX Netscaler with variable length fields should decode raw data based on cached templates
logstash_1_2681004e5afa |      Failure/Error: expect(decode[0].get("[netflow][flowEndMicroseconds]")).to eq('2016-11-11T12:09:19.000Z')
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: "2016-11-11T12:09:19.000Z"
logstash_1_2681004e5afa |             got: "2016-11-11T12:09:19.000127768Z"
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:3177:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   2) LogStash::Codecs::Netflow IPFIX Nokia BRAS should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[0].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2017-12-14T07:23:45.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"10...tocolIdentifier"=>6, "sourceIPv4Address"=>"10.0.1.228", "sourceTransportPort"=>5878, "version"=>10}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2017-12-14T07:23:45.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"10...tocolIdentifier"=>6, "sourceIPv4Address"=>"10.0.1.228", "sourceTransportPort"=>5878, "version"=>10}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |         "@timestamp" => "2017-12-14T07:23:45.000Z",
logstash_1_2681004e5afa |         "@version" => "1",
logstash_1_2681004e5afa |        -"netflow" => {"destinationIPv4Address"=>"10.0.0.34", "destinationTransportPort"=>80, "flowId"=>3389049088, "flowStartMilliseconds"=>"2017-12-14T07:23:45.148Z", "natInsideSvcid"=>100, "natOutsideSvcid"=>0, "natSubString"=>"USER1@10.10.0.123", "protocolIdentifier"=>6, "sourceIPv4Address"=>"10.0.1.228", "sourceTransportPort"=>5878, "version"=>10},
logstash_1_2681004e5afa |        +"netflow" => {"destinationIPv4Address"=>"10.0.0.34", "destinationTransportPort"=>80, "flowId"=>3389049088, "flowStartMilliseconds"=>"2017-12-14T07:23:45.148000001Z", "natInsideSvcid"=>100, "natOutsideSvcid"=>0, "natSubString"=>"USER1@10.10.0.123", "protocolIdentifier"=>6, "sourceIPv4Address"=>"10.0.1.228", "sourceTransportPort"=>5878, "version"=>10},
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:1468:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   3) LogStash::Codecs::Netflow Netflow5 mikrotik should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[29].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2016-07-21T13:51:57.514Z", "@version"=>"1", "netflow"=>{"dst_as"=>0, "dst_mask"=>0, "..., "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>40, "tcp_flags"=>82, "version"=>5}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2016-07-21T13:51:57.514932Z", "@version"=>"1", "netflow"=>{"dst_as"=>0, "dst_mask"=>0..., "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>40, "tcp_flags"=>82, "version"=>5}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |        -"@timestamp" => "2016-07-21T13:51:57.514Z",
logstash_1_2681004e5afa |        +"@timestamp" => "2016-07-21T13:51:57.514932Z",
logstash_1_2681004e5afa |         "@version" => "1",
logstash_1_2681004e5afa |        -"netflow" => {"dst_as"=>0, "dst_mask"=>0, "engine_id"=>0, "engine_type"=>0, "first_switched"=>"2016-07-21T13:51:42.254Z", "flow_records"=>30, "flow_seq_num"=>8140050, "in_bytes"=>11442, "in_pkts"=>13, "input_snmp"=>13, "ipv4_dst_addr"=>"192.168.0.1", "ipv4_next_hop"=>"192.168.0.1", "ipv4_src_addr"=>"10.0.8.1", "l4_dst_port"=>51826, "l4_src_port"=>80, "last_switched"=>"2016-07-21T13:51:42.254Z", "output_snmp"=>46, "protocol"=>6, "sampling_algorithm"=>0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>40, "tcp_flags"=>82, "version"=>5},
logstash_1_2681004e5afa |        +"netflow" => {"dst_as"=>0, "dst_mask"=>0, "engine_id"=>0, "engine_type"=>0, "first_switched"=>"2016-07-21T13:51:42.254932Z", "flow_records"=>30, "flow_seq_num"=>8140050, "in_bytes"=>11442, "in_pkts"=>13, "input_snmp"=>13, "ipv4_dst_addr"=>"192.168.0.1", "ipv4_next_hop"=>"192.168.0.1", "ipv4_src_addr"=>"10.0.8.1", "l4_dst_port"=>51826, "l4_src_port"=>80, "last_switched"=>"2016-07-21T13:51:42.254932Z", "output_snmp"=>46, "protocol"=>6, "sampling_algorithm"=>0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>40, "tcp_flags"=>82, "version"=>5},
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:884:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   4) LogStash::Codecs::Netflow IPFIX YAF basic with applabel should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[0].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2016-12-25T13:03:38.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"17...53, "sourceIPv4Address"=>"172.16.32.201", "sourceTransportPort"=>46086, "version"=>10, "vlanId"=>0}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2016-12-25T13:03:38.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"17...53, "sourceIPv4Address"=>"172.16.32.201", "sourceTransportPort"=>46086, "version"=>10, "vlanId"=>0}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |         "@timestamp" => "2016-12-25T13:03:38.000Z",
logstash_1_2681004e5afa |         "@version" => "1",
logstash_1_2681004e5afa |        -"netflow" => {"destinationIPv4Address"=>"172.16.32.100", "destinationTransportPort"=>53, "flowAttributes"=>1, "flowEndMilliseconds"=>"2016-12-25T12:58:35.819Z", "flowEndReason"=>1, "flowStartMilliseconds"=>"2016-12-25T12:58:35.818Z", "ipClassOfService"=>0, "octetTotalCount"=>132, "packetTotalCount"=>2, "protocolIdentifier"=>17, "reverseFlowAttributes"=>0, "reverseFlowDeltaMilliseconds"=>1, "reverseIpClassOfService"=>0, "reverseOctetTotalCount"=>200, "reversePacketTotalCount"=>2, "reverseVlanId"=>0, "silkAppLabel"=>53, "sourceIPv4Address"=>"172.16.32.201", "sourceTransportPort"=>46086, "version"=>10, "vlanId"=>0},
logstash_1_2681004e5afa |        +"netflow" => {"destinationIPv4Address"=>"172.16.32.100", "destinationTransportPort"=>53, "flowAttributes"=>1, "flowEndMilliseconds"=>"2016-12-25T12:58:35.819000005Z", "flowEndReason"=>1, "flowStartMilliseconds"=>"2016-12-25T12:58:35.818000078Z", "ipClassOfService"=>0, "octetTotalCount"=>132, "packetTotalCount"=>2, "protocolIdentifier"=>17, "reverseFlowAttributes"=>0, "reverseFlowDeltaMilliseconds"=>1, "reverseIpClassOfService"=>0, "reverseOctetTotalCount"=>200, "reversePacketTotalCount"=>2, "reverseVlanId"=>0, "silkAppLabel"=>53, "sourceIPv4Address"=>"172.16.32.201", "sourceTransportPort"=>46086, "version"=>10, "vlanId"=>0},
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:3019:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   5) LogStash::Codecs::Netflow IPFIX from IXIA something something should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[0].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2018-10-25T12:24:43.000Z", "@version"=>"1", "netflow"=>{"bgpDestinationAsNumber"=>240...eIPv4Address"=>"119.103.128.175", "sourceTransportPort"=>51695, "tcpControlBits"=>0, "version"=>10}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2018-10-25T12:24:43.000Z", "@version"=>"1", "netflow"=>{"bgpDestinationAsNumber"=>240...eIPv4Address"=>"119.103.128.175", "sourceTransportPort"=>51695, "tcpControlBits"=>0, "version"=>10}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |         "@timestamp" => "2018-10-25T12:24:43.000Z",
logstash_1_2681004e5afa |         "@version" => "1",
logstash_1_2681004e5afa |        -"netflow" => {"bgpDestinationAsNumber"=>24090, "bgpSourceAsNumber"=>4134, "destinationIPv4Address"=>"202.170.60.247", "destinationTransportPort"=>36197, "egressInterface"=>1, "flowEndMilliseconds"=>"2018-10-25T12:24:32.022Z", "flowEndReason"=>1, "flowStartMilliseconds"=>"2018-10-25T12:24:19.881Z", "icmpTypeCodeIPv4"=>0, "ingressInterface"=>1, "ixiaBrowserId"=>0, "ixiaBrowserName"=>"-", "ixiaDeviceId"=>0, "ixiaDeviceName"=>"unknown", "ixiaDstAsName"=>"UNISAINS-AS-AP Universiti Sains Malaysia (USM), MY", "ixiaDstLatitude"=>5.411200046539307, "ixiaDstLongitude"=>100.33540344238281, "ixiaHttpHostName"=>"", "ixiaHttpUri"=>"", "ixiaHttpUserAgent"=>"", "ixiaL7AppId"=>0, "ixiaL7AppName"=>"unknown", "ixiaLatency"=>0, "ixiaRevOctetDeltaCount"=>0, "ixiaRevPacketDeltaCount"=>0, "ixiaSrcAsName"=>"CHINANET-BACKBONE No.31,Jin-rong Street, CN", "ixiaSrcLatitude"=>30.58009910583496, "ixiaSrcLongitude"=>114.27339935302734, "ixiaThreatIPv4"=>"0.0.0.0", "ixiaThreatType"=>"", "octetDeltaCount"=>360, "packetDeltaCount"=>4, "protocolIdentifier"=>17, "reverseIcmpTypeCodeIPv4"=>0, "sourceIPv4Address"=>"119.103.128.175", "sourceTransportPort"=>51695, "tcpControlBits"=>0, "version"=>10},
logstash_1_2681004e5afa |        +"netflow" => {"bgpDestinationAsNumber"=>24090, "bgpSourceAsNumber"=>4134, "destinationIPv4Address"=>"202.170.60.247", "destinationTransportPort"=>36197, "egressInterface"=>1, "flowEndMilliseconds"=>"2018-10-25T12:24:32.022000074Z", "flowEndReason"=>1, "flowStartMilliseconds"=>"2018-10-25T12:24:19.881999969Z", "icmpTypeCodeIPv4"=>0, "ingressInterface"=>1, "ixiaBrowserId"=>0, "ixiaBrowserName"=>"-", "ixiaDeviceId"=>0, "ixiaDeviceName"=>"unknown", "ixiaDstAsName"=>"UNISAINS-AS-AP Universiti Sains Malaysia (USM), MY", "ixiaDstLatitude"=>5.411200046539307, "ixiaDstLongitude"=>100.33540344238281, "ixiaHttpHostName"=>"", "ixiaHttpUri"=>"", "ixiaHttpUserAgent"=>"", "ixiaL7AppId"=>0, "ixiaL7AppName"=>"unknown", "ixiaLatency"=>0, "ixiaRevOctetDeltaCount"=>0, "ixiaRevPacketDeltaCount"=>0, "ixiaSrcAsName"=>"CHINANET-BACKBONE No.31,Jin-rong Street, CN", "ixiaSrcLatitude"=>30.58009910583496, "ixiaSrcLongitude"=>114.27339935302734, "ixiaThreatIPv4"=>"0.0.0.0", "ixiaThreatType"=>"", "octetDeltaCount"=>360, "packetDeltaCount"=>4, "protocolIdentifier"=>17, "reverseIcmpTypeCodeIPv4"=>0, "sourceIPv4Address"=>"119.103.128.175", "sourceTransportPort"=>51695, "tcpControlBits"=>0, "version"=>10},
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:1380:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   6) LogStash::Codecs::Netflow IPFIX Netscaler with variable length fields should decode raw data
logstash_1_2681004e5afa |      Failure/Error: expect(decode[0].get("[netflow][flowEndMicroseconds]")).to eq('2016-11-11T12:09:19.000Z')
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: "2016-11-11T12:09:19.000Z"
logstash_1_2681004e5afa |             got: "2016-11-11T12:09:19.000127768Z"
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:2088:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   7) LogStash::Codecs::Netflow IPFIX Netscaler with variable length fields should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[2].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2016-11-11T12:09:19.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"10...urceIPv4Address"=>"192.168.0.1", "sourceTransportPort"=>51053, "tcpControlBits"=>24, "version"=>10}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2016-11-11T12:09:19.000Z", "@version"=>"1", "netflow"=>{"destinationIPv4Address"=>"10...urceIPv4Address"=>"192.168.0.1", "sourceTransportPort"=>51053, "tcpControlBits"=>24, "version"=>10}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |         "@timestamp" => "2016-11-11T12:09:19.000Z",
logstash_1_2681004e5afa |         "@version" => "1",
****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:2116:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   8) LogStash::Codecs::Netflow Netflow 5 valid 01 should serialize to json
logstash_1_2681004e5afa |      Failure/Error: expect(JSON.parse(decode[0].to_json)).to eq(JSON.parse(json_events[0]))
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: {"@timestamp"=>"2015-05-02T18:38:08.280Z", "@version"=>"1", "netflow"=>{"dst_as"=>0, "dst_mask"=>0, "...0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>0, "tcp_flags"=>16, "version"=>5}}
logstash_1_2681004e5afa |             got: {"@timestamp"=>"2015-05-02T18:38:08.280328Z", "@version"=>"1", "netflow"=>{"dst_as"=>0, "dst_mask"=>0...0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>0, "tcp_flags"=>16, "version"=>5}}
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        Diff:
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |        @@ -1,4 +1,4 @@
logstash_1_2681004e5afa |        -"@timestamp" => "2015-05-02T18:38:08.280Z",
logstash_1_2681004e5afa |        +"@timestamp" => "2015-05-02T18:38:08.280328Z",
logstash_1_2681004e5afa |         "@version" => "1",
logstash_1_2681004e5afa |        -"netflow" => {"dst_as"=>0, "dst_mask"=>0, "engine_id"=>0, "engine_type"=>0, "first_switched"=>"2015-06-21T11:40:52.194Z", "flow_records"=>2, "flow_seq_num"=>0, "in_bytes"=>230, "in_pkts"=>5, "input_snmp"=>0, "ipv4_dst_addr"=>"10.0.2.15", "ipv4_next_hop"=>"0.0.0.0", "ipv4_src_addr"=>"10.0.2.2", "l4_dst_port"=>22, "l4_src_port"=>54435, "last_switched"=>"2015-05-02T18:38:08.476Z", "output_snmp"=>0, "protocol"=>6, "sampling_algorithm"=>0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>0, "tcp_flags"=>16, "version"=>5},
logstash_1_2681004e5afa |        +"netflow" => {"dst_as"=>0, "dst_mask"=>0, "engine_id"=>0, "engine_type"=>0, "first_switched"=>"2015-06-21T11:40:52.194328Z", "flow_records"=>2, "flow_seq_num"=>0, "in_bytes"=>230, "in_pkts"=>5, "input_snmp"=>0, "ipv4_dst_addr"=>"10.0.2.15", "ipv4_next_hop"=>"0.0.0.0", "ipv4_src_addr"=>"10.0.2.2", "l4_dst_port"=>22, "l4_src_port"=>54435, "last_switched"=>"2015-05-02T18:38:08.476328Z", "output_snmp"=>0, "protocol"=>6, "sampling_algorithm"=>0, "sampling_interval"=>0, "src_as"=>0, "src_mask"=>0, "src_tos"=>0, "tcp_flags"=>16, "version"=>5},
logstash_1_2681004e5afa |        
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:123:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa |   9) LogStash::Codecs::Netflow configured with include_flowset_id for ipfix should decode raw data
logstash_1_2681004e5afa |      Failure/Error: expect(decode[0].get("[netflow][flowEndMicroseconds]")).to eq('2016-11-11T12:09:19.000Z')
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        expected: "2016-11-11T12:09:19.000Z"
logstash_1_2681004e5afa |             got: "2016-11-11T12:09:19.000127768Z"
logstash_1_2681004e5afa |      
logstash_1_2681004e5afa |        (compared using ==)
logstash_1_2681004e5afa |      # ./spec/codecs/netflow_spec.rb:3218:in `block in <main>'
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa | Finished in 40.44 seconds (files took 11.16 seconds to load)
logstash_1_2681004e5afa | 101 examples, 9 failures
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa | Failed examples:
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:3174 # LogStash::Codecs::Netflow configured with template caching IPFIX Netscaler with variable length fields should decode raw data based on cached templates
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:1467 # LogStash::Codecs::Netflow IPFIX Nokia BRAS should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:883 # LogStash::Codecs::Netflow Netflow5 mikrotik should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:3018 # LogStash::Codecs::Netflow IPFIX YAF basic with applabel should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:1379 # LogStash::Codecs::Netflow IPFIX from IXIA something something should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:2083 # LogStash::Codecs::Netflow IPFIX Netscaler with variable length fields should decode raw data
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:2115 # LogStash::Codecs::Netflow IPFIX Netscaler with variable length fields should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:121 # LogStash::Codecs::Netflow Netflow 5 valid 01 should serialize to json
logstash_1_2681004e5afa | rspec ./spec/codecs/netflow_spec.rb:3215 # LogStash::Codecs::Netflow configured with include_flowset_id for ipfix should decode raw data
logstash_1_2681004e5afa | 
logstash_1_2681004e5afa | Randomized with seed 16350
```

## Task to do after issue resolution
- [ ] Re-enable it in Logstash `ci/test_plugins.rb`

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

Updates the milliseconds rounding for IPFIX start/end milliseconds fields.  Fix the test to run on Logstash 8 with microseconds precision.

## What does this PR do?

Updates the tests to adapt to Logstash 8 Timestamp nanoseconds precision.
Change the way it processes milliseconds for IPFIX fields related to start/end milliseconds, switching from a truncation at milliseconds to a rounding.

## Why is it important/What is the impact to the user?

Permit to have a green CI tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

- [ ]

## How to test this PR locally

Run the tests against a Logstash 8 checkout:
```sh
export LOGSTASH_PATH=/tmp/logstash && export LOGSTASH_SOURCE=1
bundle install
bundle exec rspec
```

## Related issues

- Closes #203

### CODE REVIEW

File: lib/logstash/codecs/netflow.rb [hunk 1]
```diff
@@ -343,7 +343,13 @@ def decode_ipfix(flowset, record)
...
             case $1
             when 'Milli'
-              event[@target][k.to_s] = LogStash::Timestamp.at(v.snapshot.to_f / 1_000).to_iso8601
+              secs = v.snapshot.to_i / 1000
+              micros = (v.snapshot.to_i % 1000) * 1000
+              # Use the 2 args Timestamp.at to avoid the precision under milliseconds. Doing math division (like /1000 on a float)
+              # could introduce error in representation that makes 0.192 millis to be expressed like 0.192000001 nanoseconds,
+              # so here we cut to millis, but there is a rounding when representing to to_iso8601, so 191998 micros becomes
+              # 192 millis in LogStash 8 while in previous versions it appears truncated like 191.
+              event[@target][k.to_s] = LogStash::Timestamp.at(secs, micros).to_iso8601
```
Comment (andsel) [inline lib/logstash/codecs/netflow.rb:352]: Note for reviewers:
This force to update the test for `flow[Start|End]Milliseconds`, while in previous case when using `LogStash::Timestamp.at(milliseconds)` builder method resulted in cutting the to the 3 least significant digits, with this change is applies a rounding.
So the `191998` microseconds are rounded to `192`9 while previously was "cutted" at `191`. This is the reason why some test were updated.

The motivation for this change is that with the previous code, in LS8 Timestamp that resulted in a value with nanoseconds precision, while in LS7 was with milliseconds precision.

File: spec/codecs/netflow_spec.rb [hunk 2]
```diff
@@ -2038,7 +2050,7 @@
             "netscalerHttpReqUserAgent": "Mozilla/5.0 (Commodore 64;  kobo.com) Gecko/20100101 Firefox/75.0",
             "destinationTransportPort": 443,
             "netscalerHttpReqCookie": "beer=123456789abcdefghijklmnopqrstuvw; AnotherCookie=1234567890abcdefghijklmnopqr; Shameless.Plug=Thankyou.Rakuten.Kobo.Inc.For.Allowing.me.time.to.work.on.this.and.contribute.back.to.the.community; Padding=aaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbccccccccccccccddddddddddddddddddddddeeeeeeeeeeeeeeeeeeeeeffffffffffffffffffffffgggggggggggggggggggggggghhhhhhhhhhhhhhhhhiiiiiiiiiiiiiiiiiiiiiijjjjjjjjjjjjjjjjjjjjjjjjkkkkkkkkkkkkkkkkkklllllllllllllllmmmmmmmmmm; more=less; GJquote=There.is.no.spoon; GarrySays=Nice!!; LastPadding=aaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbcccccccccccccccccccdddddddddddeeeeeeee",
-            "flowEndMicroseconds": "2016-11-11T12:09:19.000Z",
+            "flowEndMicroseconds": "2016-11-11T12:09:19.000#{nanos}Z",
```
Comment (robbavey) [inline spec/codecs/netflow_spec.rb:2053]: Do you think it makes sense to have tests for `flowEndMicroseconds` and `flowEndNanoseconds`? I think we should at least have a test for µs precision as well as ms and ns

**Review by robbavey — COMMENTED** [review]
Small nit on adding a test for µs precision

File: spec/codecs/netflow_spec.rb (hunk 2 above)
Comment (andsel) [reply spec/codecs/netflow_spec.rb:2053]: I think that to insert that test we have to update the fixture data, in particular binary data files like:
- `ipfix_test_netscaler_tpl.dat`
- `ipfix_test_netscaler_data.dat`
which I don't know where they came from. So it's not clear to me if (and how) we can update those binary files.

**Review by robbavey — APPROVED** [review]
LGTM

**Pull request merged**